From acfdd24bd3d6d4adf1a07dcf6a0a11a6ba6ab062 Mon Sep 17 00:00:00 2001
From: Alexander Dunaev <adunaev@igalia.com>
Date: Tue, 24 May 2022 09:26:43 +0000
Subject: [PATCH] [linux/wayland] Introduced a workaround for the crash. [merge to M102]

In recent versions we have got a crash (see the linked crbug) that is
unclear how to investigate.  The code has DCHECKs but apparently users
who report crashes do not have means to diagnose the issue.

This patch adds a workaround (early return of the default) so that the
browser should fall back to the default path in the situation that
caused the crash.

(cherry picked from commit 09ccc586c172f5d1f77fa698ae73a2079e4cacc6)

Bug: 1323635
Change-Id: I1f8a0bc79cf866d9f3405c21c7c0e5b868a55bf5
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3657308
Commit-Queue: Maksim Sisov <msisov@igalia.com>
Reviewed-by: Maksim Sisov <msisov@igalia.com>
Auto-Submit: Alexander Dunaev <adunaev@igalia.com>
Commit-Queue: Alexander Dunaev <adunaev@igalia.com>
Cr-Original-Commit-Position: refs/heads/main@{#1005731}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3660189
Bot-Commit: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Commit-Queue: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/5005@{#986}
Cr-Branched-From: 5b4d9450fee01f821b6400e947b3839727643a71-refs/heads/main@{#992738}
---

--- a/ui/ozone/platform/wayland/host/wayland_window.cc
+++ b/ui/ozone/platform/wayland/host/wayland_window.cc
@@ -175,10 +175,18 @@
     auto* output_manager = connection_->wayland_output_manager();
     auto* output = output_manager->GetOutput(output_id);
     auto* preferred_output = output_manager->GetOutput(preferred_output_id);
-    // crbug.com/1323635
+    // The compositor may have told the surface to enter the output that the
+    // client is not aware of.  In such an event, we cannot evaluate scales, and
+    // can only return the default, which means falling back to the primary
+    // display in the code that calls this.
+    // DCHECKS below are kept for trying to catch the situation in developer's
+    // builds and find the way to reproduce the issue.
+    // See crbug.com/1323635
     DCHECK(output) << " output " << output_id << " not found!";
     DCHECK(preferred_output)
         << " output " << preferred_output_id << " not found!";
+    if (!output || !preferred_output)
+      return 0;
     if (output->scale_factor() > preferred_output->scale_factor())
       preferred_output_id = output_id;
   }
